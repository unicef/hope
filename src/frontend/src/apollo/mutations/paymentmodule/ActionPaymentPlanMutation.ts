import { gql } from '@apollo/client';

export const ACTION_PAYMENT_PLAN = gql`
  mutation ActionPP($input: ActionPaymentPlanInput!) {
    actionPaymentPlanMutation(input: $input) {
      paymentPlan {
        id
        status
      }
    }
  }
`;
