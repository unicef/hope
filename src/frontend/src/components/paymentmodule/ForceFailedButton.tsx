import { Box, Button, DialogContent, DialogTitle } from '@mui/material';
import * as React from 'react';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Dialog } from '@containers/dialogs/Dialog';
import { DialogActions } from '@containers/dialogs/DialogActions';
import { DialogContainer } from '@containers/dialogs/DialogContainer';
import { DialogFooter } from '@containers/dialogs/DialogFooter';
import { DialogTitleWrapper } from '@containers/dialogs/DialogTitleWrapper';
import { useSnackbar } from '@hooks/useSnackBar';
import { useMarkPayAsFailedMutation } from '@generated/graphql';

export interface ForceFailedButtonProps {
  paymentId: string;
  disabled?: boolean;
}
export function ForceFailedButton({
  paymentId,
  disabled = false,
}: ForceFailedButtonProps): React.ReactElement {
  const { t } = useTranslation();
  const [isOpenModal, setOpenModal] = useState(false);
  const { showMessage } = useSnackbar();
  const [mutate, { loading }] = useMarkPayAsFailedMutation();

  const submit = async (): Promise<void> => {
    try {
      await mutate({
        variables: {
          paymentId,
        },
      });
      setOpenModal(false);
      showMessage(t('Payment has been marked as failed.'));
    } catch (e) {
      e.graphQLErrors.map((x) => showMessage(x.message));
    }
  };

  return (
    <Box>
      <Box p={2}>
        <Button
          color="primary"
          variant="contained"
          onClick={() => setOpenModal(true)}
          data-cy="button-mark-as-failed"
          disabled={disabled}
        >
          {t('Mark as failed')}
        </Button>
      </Box>
      <Dialog
        open={isOpenModal}
        onClose={() => setOpenModal(false)}
        scroll="paper"
        aria-labelledby="form-dialog-title"
        maxWidth="md"
      >
        <DialogTitleWrapper>
          <DialogTitle>{t('Mark as failed')}</DialogTitle>
        </DialogTitleWrapper>
        <DialogContent>
          <DialogContainer>
            <Box p={5}>
              {t('Are you sure you would like to mark payment as failed?')}
            </Box>
          </DialogContainer>
        </DialogContent>
        <DialogFooter>
          <DialogActions>
            <Button onClick={() => setOpenModal(false)}>{t('CANCEL')}</Button>
            <Button
              type="submit"
              color="primary"
              variant="contained"
              onClick={() => submit()}
              data-cy="button-submit"
              disabled={loading}
            >
              {t('Mark as failed')}
            </Button>
          </DialogActions>
        </DialogFooter>
      </Dialog>
    </Box>
  );
}
