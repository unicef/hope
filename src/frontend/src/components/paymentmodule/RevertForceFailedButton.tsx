import { Box, Button, DialogContent, DialogTitle } from '@mui/material';
import * as React from 'react';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Field, Form, Formik } from 'formik';
import * as Yup from 'yup';
import CalendarTodayRoundedIcon from '@mui/icons-material/CalendarTodayRounded';
import { Dialog } from '@containers/dialogs/Dialog';
import { DialogActions } from '@containers/dialogs/DialogActions';
import { DialogFooter } from '@containers/dialogs/DialogFooter';
import { DialogTitleWrapper } from '@containers/dialogs/DialogTitleWrapper';
import { useSnackbar } from '@hooks/useSnackBar';
import { FormikTextField } from '@shared/Formik/FormikTextField';
import { FormikDateField } from '@shared/Formik/FormikDateField';
import { useRevertMarkPayAsFailedMutation } from '@generated/graphql';

export interface RevertForceFailedButtonProps {
  paymentId: string;
  disabled?: boolean;
}
export function RevertForceFailedButton({
  paymentId,
  disabled = false,
}: RevertForceFailedButtonProps): React.ReactElement {
  const { t } = useTranslation();
  const [isOpenModal, setOpenModal] = useState(false);
  const { showMessage } = useSnackbar();
  const [mutate, { loading }] = useRevertMarkPayAsFailedMutation();

  const validationSchema = Yup.object().shape({
    deliveredQuantity: Yup.number()
      .min(0)
      .max(99999999, t('Number is too big')),
    deliveryDate: Yup.date().required(t('Delivery date is required')),
  });

  const submit = async (values, { resetForm }): Promise<void> => {
    try {
      await mutate({
        variables: {
          paymentId,
          deliveredQuantity: values.deliveredQuantity,
          deliveryDate: values.deliveryDate,
        },
      });
      setOpenModal(false);
      showMessage(t('Force failed has been reverted.'));
      resetForm();
    } catch (e) {
      e.graphQLErrors.map((x) => showMessage(x.message));
    }
  };

  return (
    <Box>
      <Box p={2}>
        <Button
          color="primary"
          variant="contained"
          onClick={() => setOpenModal(true)}
          data-cy="button-revert-mark-as-failed"
          disabled={disabled}
        >
          {t('Revert mark as failed')}
        </Button>
      </Box>
      <Formik
        initialValues={{
          deliveredQuantity: 0,
          deliveryDate: '',
        }}
        validationSchema={validationSchema}
        onSubmit={submit}
      >
        {({ submitForm, resetForm }) => (
          <Dialog
            open={isOpenModal}
            onClose={() => setOpenModal(false)}
            scroll="paper"
            aria-labelledby="form-dialog-title"
            maxWidth="md"
          >
            <DialogTitleWrapper>
              <DialogTitle>{t('Revert mark as failed')}</DialogTitle>
            </DialogTitleWrapper>
            <DialogContent>
              <Form>
                <Field
                  name="deliveredQuantity"
                  label={t('Delivered Quantity')}
                  type="number"
                  required
                  fullWidth
                  variant="outlined"
                  component={FormikTextField}
                />
                <Field
                  name="deliveryDate"
                  label={t('Delivery Date')}
                  component={FormikDateField}
                  required
                  fullWidth
                  decoratorEnd={<CalendarTodayRoundedIcon color="disabled" />}
                />
              </Form>
            </DialogContent>
            <DialogFooter>
              <DialogActions>
                <Button
                  onClick={() => {
                    setOpenModal(false);
                    resetForm();
                  }}
                >
                  {t('CANCEL')}
                </Button>
                <Button
                  type="submit"
                  color="primary"
                  variant="contained"
                  onClick={submitForm}
                  data-cy="button-submit"
                  disabled={loading}
                >
                  {t('Revert mark as failed')}
                </Button>
              </DialogActions>
            </DialogFooter>
          </Dialog>
        )}
      </Formik>
    </Box>
  );
}
