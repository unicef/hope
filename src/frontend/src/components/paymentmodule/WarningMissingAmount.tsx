import { Box } from '@mui/material';
import WarningIcon from '@mui/icons-material/Warning';
import { useTranslation } from 'react-i18next';
import styled from 'styled-components';
import { ReactElement } from 'react';

const WarnIcon = styled(WarningIcon)`
  color: ${({ theme }) => theme.hctPalette.orange};
`;

interface WarningMissingAmountProps {
  amount: number;
  currency: string;
}

export function WarningMissingAmount({
  amount,
  currency,
}: WarningMissingAmountProps): ReactElement {
  const { t } = useTranslation();

  return (
    <Box display="flex" alignItems="center">
      <WarnIcon />
      {t('Missing')} {amount} {currency}
    </Box>
  );
}
