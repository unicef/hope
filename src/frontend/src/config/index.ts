export const API_BASE_URL = '/api';
export const GRAPHQL_URL = `${API_BASE_URL}/graphql`;
export const LOGIN_URL = '/api/login/azuread-tenant-oauth2/';
export const DATE_FORMAT = 'D MMM YYYY';
export const DATE_FORMAT_LONG = 'D MMM YYYY h:mm A';
export const AUTO_LOGOUT_MINUTES = 15;
export const AUTO_LOGOUT_MILLIS = 1000 * 60 * AUTO_LOGOUT_MINUTES;
