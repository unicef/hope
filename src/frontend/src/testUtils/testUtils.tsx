import { ReactElement } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render, RenderOptions } from '@testing-library/react';
import { Formik } from 'formik';
import noop from 'lodash/noop';
import { TestProviders } from './testProviders';
import { ProgramStatus } from '@generated/graphql';
import { MockLink } from '@apollo/client/testing';

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) =>
  render(
    <>
      <BrowserRouter>
        <Formik initialValues={{}} onSubmit={noop}>
          {ui}
        </Formik>
      </BrowserRouter>
    </>,
    {
      wrapper: TestProviders,
      ...options,
    },
  );

export * from '@testing-library/react';
export { customRender as render };

let m_w = 123456789;
let m_z = 987654321;
const mask = 0xffffffff;

// Takes any integer
export const seed = (i) => {
  m_w = (123456789 + i) & mask;
  m_z = (987654321 - i) & mask;
};

// Returns number between 0 (inclusive) and 1.0 (exclusive),
// just like Math.random().
export const random = () => {
  m_z = (36969 * (m_z & 65535) + (m_z >> 16)) & mask;
  m_w = (18000 * (m_w & 65535) + (m_w >> 16)) & mask;
  let result = ((m_z << 16) + (m_w & 65535)) >>> 0;
  result /= 4294967296;
  return result;
};

export class ApolloLoadingLink extends MockLink {
  constructor() {
    super([]);
  }

  request() {
    return null;
  }
}

export const fakeContextProgram = {
  selectedProgram: {
    id: '1',
    name: 'someName',
    status: ProgramStatus.Active,
    pduFields: null,
    dataCollectingType: {
      id: '1',
      householdFiltersAvailable: true,
      individualFiltersAvailable: true,
      label: 'data collecting type',
      code: '123',
      type: 'full',
      children: null,
    },
  },
  setSelectedProgram: () => {},
  isActiveProgram: true,
  isSocialDctType: false,
  isStandardDctType: true,
  programHasPdu: false,
};
