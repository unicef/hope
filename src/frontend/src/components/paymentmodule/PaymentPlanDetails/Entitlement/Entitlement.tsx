import {
  Box,
  Button,
  FormControl,
  Grid,
  InputLabel,
  MenuItem,
  Select,
  Typography,
} from '@mui/material';
import { GetApp } from '@mui/icons-material';
import AttachFileIcon from '@mui/icons-material/AttachFile';
import * as React from 'react';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import styled from 'styled-components';
import { hasPermissions, PERMISSIONS } from '../../../../config/permissions';
import { useSnackbar } from '@hooks/useSnackBar';
import {
  PaymentPlanBackgroundActionStatus,
  PaymentPlanDocument,
  PaymentPlanQuery,
  PaymentPlanStatus,
  useAllSteficonRulesQuery,
  useExportXlsxPpListMutation,
  useSetSteficonRuleOnPpListMutation,
} from '@generated/graphql';
import { ContainerColumnWithBorder } from '@core/ContainerColumnWithBorder';
import { LabelizedField } from '@core/LabelizedField';
import { LoadingButton } from '@core/LoadingButton';
import { LoadingComponent } from '@core/LoadingComponent';
import { Title } from '@core/Title';
import { UniversalMoment } from '@core/UniversalMoment';
import { BigValue } from '../../../rdi/details/RegistrationDetails/RegistrationDetails';
import { ImportXlsxPaymentPlanPaymentList } from '../ImportXlsxPaymentPlanPaymentList/ImportXlsxPaymentPlanPaymentList';
import { useProgramContext } from '../../../../programContext';

const GreyText = styled.p`
  color: #9e9e9e;
  font-size: 16px;
`;

const GreyTextSmall = styled.p`
  color: #9e9e9e;
  font-size: 14px;
`;

const OrDivider = styled.div`
  border-top: 1px solid #b1b1b5;
  height: 2px;
  width: 50%;
  margin-top: 20px;
`;

const Divider = styled.div`
  border-top: 1px solid #b1b1b5;
  height: 20px;
`;

const DividerLabel = styled.div`
  width: 40px;
  height: 40px;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 14px;
  font-weight: 500;
  color: #253b46;
  text-transform: uppercase;
  padding: 5px;
  border: 1px solid #b1b1b5;
  border-radius: 50%;
  background-color: #fff;
  margin-top: 20px;
`;

const DownloadIcon = styled(GetApp)`
  color: #043f91;
`;

const SpinaczIconContainer = styled(Box)`
  position: relative;
  top: 4px;
  font-size: 16px;
  color: #666666;
`;

const BoxWithBorderRight = styled(Box)`
  border-right: 1px solid #b1b1b5;
`;

interface EntitlementProps {
  paymentPlan: PaymentPlanQuery['paymentPlan'];
  permissions: string[];
}

export function Entitlement({
  paymentPlan,
  permissions,
}: EntitlementProps): React.ReactElement {
  const { t } = useTranslation();
  const { showMessage } = useSnackbar();
  const { isActiveProgram } = useProgramContext();

  const [steficonRuleValue, setSteficonRuleValue] = useState<string>(
    paymentPlan.steficonRule?.rule.id || '',
  );
  const options = {
    refetchQueries: () => [
      {
        query: PaymentPlanDocument,
        variables: {
          id: paymentPlan.id,
        },
      },
    ],
  };

  const [setSteficonRule, { loading: loadingSetSteficonRule }] =
    useSetSteficonRuleOnPpListMutation(options);

  const { data: steficonData, loading } = useAllSteficonRulesQuery({
    variables: { enabled: true, deprecated: false, type: 'PAYMENT_PLAN' },
    fetchPolicy: 'network-only',
  });
  const [mutateExport, { loading: loadingExport }] =
    useExportXlsxPpListMutation();

  if (!steficonData) {
    return null;
  }
  if (loading) {
    return <LoadingComponent />;
  }

  const canApplySteficonRule = hasPermissions(
    PERMISSIONS.PM_APPLY_RULE_ENGINE_FORMULA_WITH_ENTITLEMENTS,
    permissions,
  );

  const shouldDisableEntitlementSelect =
    !canApplySteficonRule ||
    paymentPlan.status !== PaymentPlanStatus.Locked ||
    !isActiveProgram;

  const shouldDisableDownloadTemplate =
    paymentPlan.status !== PaymentPlanStatus.Locked || !isActiveProgram;

  const shouldDisableExportXlsx =
    loadingExport ||
    paymentPlan.status !== PaymentPlanStatus.Locked ||
    paymentPlan?.backgroundActionStatus ===
      PaymentPlanBackgroundActionStatus.XlsxExporting ||
    !isActiveProgram;

  return (
    <Box m={5}>
      <ContainerColumnWithBorder>
        <Box mt={4}>
          <Title>
            <Typography variant="h6">{t('Entitlement')}</Typography>
          </Title>
          <GreyText>{t('Select Entitlement Formula')}</GreyText>
          <Grid alignItems="center" container>
            <Grid item xs={11}>
              <FormControl size="small" variant="outlined" fullWidth>
                <Box mb={1}>
                  <InputLabel>{t('Entitlement Formula')}</InputLabel>
                </Box>
                <Select
                  size="small"
                  disabled={shouldDisableEntitlementSelect}
                  MenuProps={{
                    anchorOrigin: {
                      vertical: 'bottom',
                      horizontal: 'left',
                    },
                    transformOrigin: {
                      vertical: 'top',
                      horizontal: 'left',
                    },
                  }}
                  value={steficonRuleValue}
                  data-cy="input-entitlement-formula"
                  onChange={(event) => setSteficonRuleValue(event.target.value)}
                >
                  {steficonData.allSteficonRules?.edges?.map((each, index) => (
                    <MenuItem
                      data-cy={`select-option-${index}`}
                      key={each.node.id}
                      value={each.node.id}
                    >
                      {each.node.name}
                    </MenuItem>
                  ))}
                </Select>
              </FormControl>
            </Grid>
            <Grid item>
              <Box ml={2}>
                <Button
                  variant="contained"
                  color="primary"
                  disabled={
                    loadingSetSteficonRule ||
                    !steficonRuleValue ||
                    paymentPlan.status !== PaymentPlanStatus.Locked ||
                    paymentPlan.backgroundActionStatus ===
                      PaymentPlanBackgroundActionStatus.RuleEngineRun ||
                    !isActiveProgram
                  }
                  data-cy="button-apply-steficon"
                  onClick={async () => {
                    try {
                      await setSteficonRule({
                        variables: {
                          paymentPlanId: paymentPlan.id,
                          steficonRuleId: steficonRuleValue,
                        },
                      });
                      showMessage(
                        t('Formula is executing, please wait until completed'),
                      );
                    } catch (e) {
                      e.graphQLErrors.map((x) => showMessage(x.message));
                    }
                  }}
                >
                  {t('Apply')}
                </Button>
              </Box>
            </Grid>
          </Grid>
          <Box display="flex" alignItems="center">
            <OrDivider />
            <DividerLabel>Or</DividerLabel>
            <OrDivider />
          </Box>
        </Box>
        <Box display="flex">
          <Box width="50%">
            <BoxWithBorderRight
              display="flex"
              justifyContent="center"
              alignItems="center"
              flexDirection="column"
            >
              {paymentPlan.hasPaymentListExportFile ? (
                <Button
                  color="primary"
                  startIcon={<DownloadIcon />}
                  component="a"
                  download
                  data-cy="button-download-template"
                  href={`/api/download-payment-plan-payment-list/${paymentPlan.id}`}
                  disabled={shouldDisableDownloadTemplate}
                >
                  {t('DOWNLOAD TEMPLATE')}
                </Button>
              ) : (
                <LoadingButton
                  loading={loadingExport}
                  disabled={shouldDisableExportXlsx}
                  color="primary"
                  startIcon={<GetApp />}
                  data-cy="button-export-xlsx"
                  onClick={async () => {
                    try {
                      await mutateExport({
                        variables: {
                          paymentPlanId: paymentPlan.id,
                        },
                      });
                      showMessage(
                        t('Exporting XLSX started. Please check your email.'),
                      );
                    } catch (e) {
                      e.graphQLErrors.map((x) => showMessage(x.message));
                    }
                  }}
                >
                  {t('Export Xlsx')}
                </LoadingButton>
              )}

              <GreyTextSmall>
                {t(
                  'Template contains payment list with all targeted households',
                )}
              </GreyTextSmall>
            </BoxWithBorderRight>
          </Box>
          <Box width="50%">
            <Box
              display="flex"
              justifyContent="center"
              alignItems="center"
              flexDirection="column"
            >
              <Box>
                <ImportXlsxPaymentPlanPaymentList
                  permissions={permissions}
                  paymentPlan={paymentPlan}
                />
              </Box>
              {paymentPlan?.importedFileName ? (
                <Box alignItems="center" display="flex">
                  <SpinaczIconContainer>
                    <AttachFileIcon fontSize="inherit" />
                  </SpinaczIconContainer>
                  <Box mr={1}>
                    <GreyTextSmall data-cy="imported-file-name">
                      {paymentPlan?.importedFileName}
                    </GreyTextSmall>
                  </Box>
                  <GreyTextSmall>
                    {paymentPlan?.importedFileDate ? (
                      <UniversalMoment>
                        {paymentPlan?.importedFileDate}
                      </UniversalMoment>
                    ) : null}
                  </GreyTextSmall>
                </Box>
              ) : (
                <GreyTextSmall>
                  {t(
                    'Uploaded file should contain entitlement for each household',
                  )}
                </GreyTextSmall>
              )}
            </Box>
          </Box>
        </Box>
        {paymentPlan.totalEntitledQuantityUsd ? (
          <>
            <Divider />
            <LabelizedField label={t('Total Entitled Quantity')}>
              <BigValue data-cy="total-entitled-quantity-usd">
                {`${paymentPlan.totalEntitledQuantity} ${paymentPlan.currency} (${paymentPlan.totalEntitledQuantityUsd} USD)`}
              </BigValue>
            </LabelizedField>
          </>
        ) : null}
      </ContainerColumnWithBorder>
    </Box>
  );
}
