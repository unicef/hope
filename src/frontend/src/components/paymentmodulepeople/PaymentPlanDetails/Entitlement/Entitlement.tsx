import { ContainerColumnWithBorder } from '@core/ContainerColumnWithBorder';
import { LabelizedField } from '@core/LabelizedField';
import { LoadingButton } from '@core/LoadingButton';
import { LoadingComponent } from '@core/LoadingComponent';
import { Title } from '@core/Title';
import { UniversalMoment } from '@core/UniversalMoment';
import { PaymentPlanStatusEnum } from '@restgenerated/models/PaymentPlanStatusEnum';
import { BackgroundActionStatusEnum } from '@restgenerated/models/BackgroundActionStatusEnum';
import { useBaseUrl } from '@hooks/useBaseUrl';
import { useSnackbar } from '@hooks/useSnackBar';
import { GetApp } from '@mui/icons-material';
import AttachFileIcon from '@mui/icons-material/AttachFile';
import {
  Box,
  Button,
  FormControl,
  Grid2 as Grid,
  InputLabel,
  MenuItem,
  Select,
  Typography,
} from '@mui/material';
import { PaginatedRuleList } from '@restgenerated/models/PaginatedRuleList';
import { PaymentPlanDetail } from '@restgenerated/models/PaymentPlanDetail';
import { RestService } from '@restgenerated/services/RestService';
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { ReactElement, useState } from 'react';
import { useTranslation } from 'react-i18next';
import styled from 'styled-components';
import { hasPermissions, PERMISSIONS } from '../../../../config/permissions';
import { useProgramContext } from '../../../../programContext';
import { BigValue } from '../../../rdi/details/RegistrationDetails/RegistrationDetails';
import { ImportXlsxPaymentPlanPaymentList } from '../ImportXlsxPaymentPlanPaymentList/ImportXlsxPaymentPlanPaymentList';
import { ApplyEngineFormula } from '@restgenerated/models/ApplyEngineFormula';
import { showApiErrorMessages } from '@utils/utils';

const GreyText = styled.p`
  color: #9e9e9e;
  font-size: 16px;
`;

const GreyTextSmall = styled.p`
  color: #9e9e9e;
  font-size: 14px;
`;

const OrDivider = styled.div`
  border-top: 1px solid #b1b1b5;
  height: 2px;
  width: 50%;
  margin-top: 20px;
`;

const Divider = styled.div`
  border-top: 1px solid #b1b1b5;
  height: 20px;
`;

const DividerLabel = styled.div`
  width: 40px;
  height: 40px;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 14px;
  font-weight: 500;
  color: #253b46;
  text-transform: uppercase;
  padding: 5px;
  border: 1px solid #b1b1b5;
  border-radius: 50%;
  background-color: #fff;
  margin-top: 20px;
`;

const DownloadIcon = styled(GetApp)`
  color: #043f91;
`;

const SpinaczIconContainer = styled(Box)`
  position: relative;
  top: 4px;
  font-size: 16px;
  color: #666666;
`;

const BoxWithBorderRight = styled(Box)`
  border-right: 1px solid #b1b1b5;
`;

interface EntitlementProps {
  paymentPlan: PaymentPlanDetail;
  permissions: string[];
}

export function Entitlement({
  paymentPlan,
  permissions,
}: EntitlementProps): ReactElement {
  const { t } = useTranslation();
  const { showMessage } = useSnackbar();
  const { isActiveProgram } = useProgramContext();
  const { businessArea, programId } = useBaseUrl();
  const queryClient = useQueryClient();

  const [steficonRuleValue, setSteficonRuleValue] = useState<string>(
    paymentPlan.steficonRule?.id ? String(paymentPlan.steficonRule.id) : '',
  );

  const { mutateAsync: setSteficonRule, isPending: loadingSetSteficonRule } =
    useMutation({
      mutationFn: ({
        businessAreaSlug,
        id,
        programSlug,
        requestBody,
      }: {
        businessAreaSlug: string;
        id: string;
        programSlug: string;
        requestBody: ApplyEngineFormula;
      }) =>
        RestService.restBusinessAreasProgramsPaymentPlansApplyEngineFormulaCreate(
          {
            businessAreaSlug,
            id,
            programSlug,
            requestBody,
          },
        ),
      onSuccess: () => {
        showMessage(t('Formula is executing, please wait until completed'));
        queryClient.invalidateQueries({
          queryKey: ['paymentPlan', businessArea, paymentPlan.id, programId],
        });
      },
      onError: (e) => {
        showMessage(e.message);
      },
    });

  const { data: steficonData, isLoading: loading } =
    useQuery<PaginatedRuleList>({
      queryKey: ['engineRules'],
      queryFn: () =>
        RestService.restEngineRulesList({
          type: 'PAYMENT_PLAN',
          deprecated: false,
          enabled: true,
        }),
    });

  const { mutateAsync: mutateExport, isPending: loadingExport } = useMutation({
    mutationFn: ({
      businessAreaSlug,
      id,
      programSlug,
    }: {
      businessAreaSlug: string;
      id: string;
      programSlug: string;
    }) =>
      RestService.restBusinessAreasProgramsPaymentPlansReconciliationExportXlsxRetrieve(
        {
          businessAreaSlug,
          id,
          programSlug,
        },
      ),
    onSuccess: () => {
      showMessage(t('Exporting XLSX started. Please check your email.'));
    },
    onError: (e) => {
      showMessage(e.message);
    },
  });

  if (!steficonData) {
    return null;
  }
  if (loading) {
    return <LoadingComponent />;
  }

  const canApplySteficonRule = hasPermissions(
    PERMISSIONS.PM_APPLY_RULE_ENGINE_FORMULA_WITH_ENTITLEMENTS,
    permissions,
  );

  const shouldDisableEntitlementSelect =
    !canApplySteficonRule ||
    paymentPlan.status !== PaymentPlanStatusEnum.LOCKED ||
    !isActiveProgram;

  const shouldDisableDownloadTemplate =
    paymentPlan.status !== PaymentPlanStatusEnum.LOCKED || !isActiveProgram;

  const shouldDisableExportXlsx =
    loadingExport ||
    paymentPlan.status !== PaymentPlanStatusEnum.LOCKED ||
    paymentPlan?.backgroundActionStatus ===
    BackgroundActionStatusEnum.XLSX_EXPORTING ||
    !isActiveProgram;

  return (
    <Box m={5}>
      <ContainerColumnWithBorder>
        <Box mt={4}>
          <Title>
            <Typography variant="h6">{t('Entitlement')}</Typography>
          </Title>
          <GreyText>{t('Select Entitlement Formula')}</GreyText>
          <Grid alignItems="center" container>
            <Grid size={{ xs: 11 }}>
              <FormControl size="small" variant="outlined" fullWidth>
                <Box mb={1}>
                  <InputLabel>{t('Entitlement Formula')}</InputLabel>
                </Box>
                <Select
                  size="small"
                  disabled={shouldDisableEntitlementSelect}
                  MenuProps={{
                    anchorOrigin: {
                      vertical: 'bottom',
                      horizontal: 'left',
                    },
                    transformOrigin: {
                      vertical: 'top',
                      horizontal: 'left',
                    },
                  }}
                  value={steficonRuleValue}
                  data-cy="input-entitlement-formula"
                  onChange={(event) => setSteficonRuleValue(event.target.value)}
                >
                  {steficonData?.results?.map((each, index) => (
                    <MenuItem
                      data-cy={`select-option-${index}`}
                      key={each.id}
                      value={each.id}
                    >
                      {each.name}
                    </MenuItem>
                  ))}
                </Select>
              </FormControl>
            </Grid>
            <Grid>
              <Box ml={2}>
                <Button
                  variant="contained"
                  color="primary"
                  disabled={
                    loadingSetSteficonRule ||
                    !steficonRuleValue ||
                    paymentPlan.status !== PaymentPlanStatusEnum.LOCKED ||
                    paymentPlan.backgroundActionStatus ===
                    BackgroundActionStatusEnum.RULE_ENGINE_RUN ||
                    !isActiveProgram
                  }
                  data-cy="button-apply-steficon"
                  onClick={async() => {
                    try {
                      await setSteficonRule({
                        programSlug: programId,
                        businessAreaSlug: businessArea,
                        id: paymentPlan.id,
                        requestBody: {
                          engineFormulaRuleId: steficonRuleValue,
                          version: paymentPlan.version,
                        },
                      });
                      showMessage(
                        t('Formula is executing, please wait until completed'),
                      );
                    } catch (e) {
                      showApiErrorMessages(e, showMessage);
                    }
                  }}
                >
                  {t('Apply')}
                </Button>
              </Box>
            </Grid>
          </Grid>
          <Box display="flex" alignItems="center">
            <OrDivider />
            <DividerLabel>Or</DividerLabel>
            <OrDivider />
          </Box>
        </Box>
        <Box display="flex">
          <Box width="50%">
            <BoxWithBorderRight
              display="flex"
              justifyContent="center"
              alignItems="center"
              flexDirection="column"
            >
              {paymentPlan.hasPaymentListExportFile ? (
                <Button
                  color="primary"
                  startIcon={<DownloadIcon />}
                  component="a"
                  download
                  data-cy="button-download-template"
                  href={`/api/download-payment-plan-payment-list/${paymentPlan.id}`}
                  disabled={shouldDisableDownloadTemplate}
                >
                  {t('DOWNLOAD TEMPLATE')}
                </Button>
              ) : (
                <LoadingButton
                  loading={loadingExport}
                  disabled={shouldDisableExportXlsx}
                  color="primary"
                  startIcon={<GetApp />}
                  data-cy="button-export-xlsx"
                  onClick={() =>
                    mutateExport({
                      businessAreaSlug: businessArea,
                      programSlug: programId,
                      id: paymentPlan.id,
                    })
                  }
                >
                  {t('Export Xlsx')}
                </LoadingButton>
              )}

              <GreyTextSmall>
                {t(
                  'Template contains payment list with all targeted households',
                )}
              </GreyTextSmall>
            </BoxWithBorderRight>
          </Box>
          <Box width="50%">
            <Box
              display="flex"
              justifyContent="center"
              alignItems="center"
              flexDirection="column"
            >
              <Box>
                <ImportXlsxPaymentPlanPaymentList
                  permissions={permissions}
                  paymentPlan={paymentPlan}
                />
              </Box>
              {paymentPlan?.importedFileName ? (
                <Box alignItems="center" display="flex">
                  <SpinaczIconContainer>
                    <AttachFileIcon fontSize="inherit" />
                  </SpinaczIconContainer>
                  <Box mr={1}>
                    <GreyTextSmall data-cy="imported-file-name">
                      {paymentPlan?.importedFileName}
                    </GreyTextSmall>
                  </Box>
                  <GreyTextSmall>
                    {paymentPlan?.importedFileName ? (
                      <UniversalMoment>
                        {paymentPlan?.importedFileDate}
                      </UniversalMoment>
                    ) : null}
                  </GreyTextSmall>
                </Box>
              ) : (
                <GreyTextSmall>
                  {t(
                    'Uploaded file should contain entitlement for each household',
                  )}
                </GreyTextSmall>
              )}
            </Box>
          </Box>
        </Box>
        {paymentPlan.totalEntitledQuantityUsd ? (
          <>
            <Divider />
            <LabelizedField label={t('Total Entitled Quantity')}>
              <BigValue data-cy="total-entitled-quantity-usd">
                {`${paymentPlan.totalEntitledQuantity} ${paymentPlan.currency} (${paymentPlan.totalEntitledQuantityUsd} USD)`}
              </BigValue>
            </LabelizedField>
          </>
        ) : null}
      </ContainerColumnWithBorder>
    </Box>
  );
}
