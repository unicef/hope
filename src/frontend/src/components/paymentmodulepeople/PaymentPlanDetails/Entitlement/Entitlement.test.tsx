import * as React from 'react';
import { MockedProvider } from '@apollo/react-testing';
import { act } from 'react-dom/test-utils';
import wait from 'waait';
import { fakeApolloAllSteficonRules } from '../../../../../fixtures/steficon/fakeApolloAllSteficonRules';
import { fakeApolloPaymentPlan } from '../../../../../fixtures/paymentmodule/fakeApolloPaymentPlan';
import { render } from '../../../../testUtils/testUtils';
import { Entitlement } from './Entitlement';
import {PERMISSIONS} from "../../../../config/permissions";

describe('components/paymentmodule/PaymentPlanDetails/Entitlement', () => {
  it('should render', async () => {
    const { container } = render(
      <MockedProvider addTypename={false} mocks={fakeApolloAllSteficonRules}>
        <Entitlement
          paymentPlan={fakeApolloPaymentPlan}
          permissions={[PERMISSIONS.PM_APPLY_RULE_ENGINE_FORMULA_WITH_ENTITLEMENTS]}
        />
      </MockedProvider>,
    );
    await act(() => wait(0)); // wait for response
    expect(container).toMatchSnapshot();
  });
});
