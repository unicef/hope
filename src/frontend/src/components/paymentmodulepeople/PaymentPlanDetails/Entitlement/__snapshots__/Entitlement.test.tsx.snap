// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/paymentmodule/PaymentPlanDetails/Entitlement should render 1`] = `<div />`;
